bgpd/bgp_fsm.c

```diff
@@ -1181,6 +1181,19 @@ int bgp_stop(struct peer *peer)
 		}
 	}
 
+	/* If a BGP speaker decides to de-configure a peer, then
+	 * the speaker SHOULD send a NOTIFICATION message with the
+	 * Error Code Cease and the Error Subcode "Peer De-configured".
+	 *
+	 * Make sure this is sent only after all the timers are stopped
+	 * and before the peer's socket is closed to avoid race condition
+	 * when you send a notification, FSM Connect fires before
+	 * closing the socket and you still see `Peer closed the session`.
+	 */
+	if (BGP_IS_VALID_STATE_FOR_NOTIF(peer->status))
```

Comment:
Sending notification inside bgp_stop() would lead to generating a "BGP_Stop" event inside the function bgp_write_notify(), there by calling bgp_stop() again ?

BGP_XYZ_State --> Idle ----> Idle

This can be solved by moving the bgp stop event generation from bgp_write_notify() (--> BGP_EVENT_ADD(peer, BGP_Stop) )to the respective calling function.